bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,392 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+use s2n_tls_sys::*;
+
+use crate::error::{Error, Fallible};
+
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+// this is the size of the MD5 hash digest that is used for the JA3 fingerprint
+const MD5_HASH_SIZE: u32 = 16;
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+// ClientHello is an opaque wrapper struct around `s2n_client_hello`. Note that
+// the size of this type is not known, and as such it can only be used through
+// references and pointers.
+//
+// This implementation is motivated by the different memory_management required
+// for different s2n_client_hello pointers. `s2n_client_hello_parse_message`
+// returns a `*mut s2n_client_hello` which owns it's own data. This neatly fits
+//  the "smart pointer" pattern and can be represented as a `Box<T>`.
+//
+// `s2n_connection_get_client_hello` returns a `*mut s2n_client_hello` which
+// references memory owned by the connection, and therefore must not outlive
+// the connection struct. This is best represented as a reference tied to the
+// lifetime of the `Connection` struct.
+
+/// ```no_run
```

Comment:
What's this example trying to show? And why mark it "no_run"?

Reply:
I think the String capacity thing is a pretty uncommon pattern in Rust, even though it makes sense for our use case. It seemed good to document exactly what that looks like. 

It's marked `no_run` because the code should compile, but won't actually run. Getting it to a runnable state would require either doing a handshake in the doc test or including the raw client hello bytes to parse a client hello.